world/wguild_mgr.cpp

```diff
@@ -133,15 +144,323 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+				auto r = z.get();
+				if (r->GetZoneID() > 0) {
+					r->SendPacket(pack);
+				}
+			}
+		}
+		else {
+			LogError("World Received ServerOP_GuildPermissionUpdate for guild [{}] function id {} with value of {} but guild could not be found.",
+				sg->guild_id,
+				sg->function_id,
+				sg->function_value
+			);
+		}
 
+		break;
+	}
+	case ServerOP_GuildRankNameChange:
+	{
+		if (pack->size != sizeof(ServerGuildRankNameChange))
+		{
+			LogGuilds("Received ServerOP_ServerGuildRankNameChange of incorrect size [{}], expected [{}]", pack->size, sizeof(ServerGuildPermissionUpdate_Struct));
+			return;
+		}
+
+		ServerGuildRankNameChange* rnc = (ServerGuildRankNameChange*)pack->pBuffer;
+
+		auto guild = GetGuildByGuildID(rnc->guild_id);
+		if (!guild) {
+			guild_mgr.LoadGuilds();
+			guild = GetGuildByGuildID(rnc->guild_id);
+		}
+		if (guild) {
+			guild->rank_names[rnc->rank] = rnc->rank_name;
+			LogGuilds("World Received ServerOP_GuildRankNameChange from zone for guild [{}] rank id {} with new name of {}",
+				rnc->guild_id,
+				rnc->rank,
+				rnc->rank_name
+			);
+			for (auto const& z : zoneserver_list.getZoneServerList()) {
+				auto r = z.get();
+				if (r->GetZoneID() > 0) {
+					r->SendPacket(pack);
+				}
+			}
+		}
+		else {
+			LogError("World Received ServerOP_GuildRankNameChange from zone for guild [{}] rank id {} with new name of {} but could not find guild.",
+				rnc->guild_id,
+				rnc->rank,
+				rnc->rank_name
+			);
+		}
+
+		break;
+	}
+	case ServerOP_GuildMemberLevelUpdate:
+	case ServerOP_GuildMemberPublicNote:
+	case ServerOP_GuildChannel:
+	case ServerOP_GuildURL:
+	case ServerOP_GuildMemberRemove:
+	case ServerOP_GuildMemberAdd:
+	case ServerOP_GuildSendGuildList:
+	case ServerOP_GuildMembersList:
+	{
+		for (auto const& z : zoneserver_list.getZoneServerList()) {
+			auto r = z.get();
+			if (r->GetZoneID() > 0) {
+				r->SendPacket(pack);
+			}
+		}
+		break;
+	}
 	default:
 		LogGuilds("Unknown packet {:#04x} received from zone??", pack->opcode);
 		break;
 	}
 }
 
+void WorldGuildManager::Process()
+{
+	for (auto &g: m_guilds) {
+		if (!g.second->tribute.enabled) {
+			continue;
+		}
+		else if (g.second->tribute.enabled && !g.second->tribute.timer.Enabled()) {
+			g.second->tribute.timer.Start(g.second->tribute.time_remaining);
+			LogGuilds(
+				"Found a Guild Tribute Timer for guild [{}]. that was not started.  Started it with {} time remaining before restart.",
+				g.first,
+				g.second->tribute.time_remaining
+			);
+		}
+		else if (g.second->tribute.enabled &&
+				 g.second->tribute.timer.Enabled() &&
+				 g.second->tribute.timer.Check()
+			) {
+			g.second->tribute.favor -= GetGuildTributeCost(g.first);
+			g.second->tribute.time_remaining = RuleI(Guild, TributeTime);
+			g.second->tribute.timer.Start(RuleI(Guild, TributeTime));
+
+			guild_mgr.UpdateDbGuildFavor(g.first, g.second->tribute.favor);
+			guild_mgr.UpdateDbTributeTimeRemaining(g.first, RuleI(Guild, TributeTime));
```

Comment:
Great question.  Currently it will hit the db when the Guild Tribute timer expires (every 10 mins) or when the RefreshInterval (default is 3 mins) expires.  Let me know what you think.

Reply:
If the interval is that slow then it should be fine here. It can be changed in a followup if it proves to be an issue on servers with lots of guilds.